pkg/markers: index out of range panic on collector.go:293

https://github.com/kubernetes-sigs/controller-tools/blob/4a32e77de72f35527a03dbd5813a14441d04cfbf/pkg/markers/collect.go#L293

```
$ make manifests
/Users/abursavich/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./api/...;./controllers/..." output:crd:artifacts:config=config/crd/bases
panic: runtime error: index out of range

goroutine 1 [running]:
sigs.k8s.io/controller-tools/pkg/markers.markerSubVisitor.Visit(0xc0020f7200, 0x188a100, 0xc002b4b100, 0xc002b6b000, 0x188a540, 0xc001bc9b00, 0x0, 0x0)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/markers/collect.go:293 +0x18ad
go/ast.Walk(0x1886de0, 0xc002b6b0c0, 0x188a540, 0xc001bc9b00)
	/usr/local/go/src/go/ast/walk.go:52 +0x66
go/ast.Walk(0x1886de0, 0xc002b6b0c0, 0x188a100, 0xc002b4b100)
	/usr/local/go/src/go/ast/walk.go:74 +0xba6
go/ast.Walk(0x1886de0, 0xc002b6af40, 0x188a140, 0xc001bc9b60)
	/usr/local/go/src/go/ast/walk.go:84 +0x144
go/ast.Walk(0x1886de0, 0xc002b6af20, 0x188a7c0, 0xc002b4c7e0)
	/usr/local/go/src/go/ast/walk.go:161 +0x27c8
go/ast.Walk(0x1886de0, 0xc002b6aee0, 0x188a880, 0xc001bc9980)
	/usr/local/go/src/go/ast/walk.go:318 +0x2af5
go/ast.Walk(0x1886de0, 0xc002b6aec0, 0x188a2c0, 0xc002b4b140)
	/usr/local/go/src/go/ast/walk.go:331 +0xf4a
go/ast.walkDeclList(0x1886de0, 0xc002b53740, 0xc000965e00, 0x1b, 0x20)
	/usr/local/go/src/go/ast/walk.go:38 +0x9e
go/ast.Walk(0x1886de0, 0xc002b53740, 0x188a180, 0xc00136db00)
	/usr/local/go/src/go/ast/walk.go:353 +0x2659
sigs.k8s.io/controller-tools/pkg/markers.(*Collector).associateFileMarkers(0xc000245f80, 0xc00136db00, 0xc00022a000)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/markers/collect.go:156 +0x10e
sigs.k8s.io/controller-tools/pkg/markers.(*Collector).associatePkgMarkers(0xc000245f80, 0xc000245f60, 0xc0007f0240)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/markers/collect.go:137 +0x27c
sigs.k8s.io/controller-tools/pkg/markers.(*Collector).MarkersInPackage(0xc000245f80, 0xc000245f60, 0x0, 0x0, 0x0)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/markers/collect.go:86 +0xe1
sigs.k8s.io/controller-tools/pkg/markers.PackageMarkers(0xc000245f80, 0xc000245f60, 0xc0001cf3c0, 0x10430ec, 0x10)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/markers/zip.go:81 +0x50
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).indexTypes(0xc0000ea0a0, 0xc000245f60)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/crd/parser.go:103 +0x3c
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).AddPackage(0xc0000ea0a0, 0xc000245f60)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/crd/parser.go:172 +0xc0
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedPackage(0xc0000ea0a0, 0xc000245f60)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/crd/parser.go:191 +0x1af
sigs.k8s.io/controller-tools/pkg/crd.Generator.Generate(0x4e00000000000001, 0xc0000ea000, 0x1d5dc40, 0xc00019e6c0)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/crd/gen.go:51 +0xce
sigs.k8s.io/controller-tools/pkg/genall.(*Runtime).Run(0xc000816100, 0x3)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/pkg/genall/genall.go:156 +0x170
main.runGenerators(0xc000170550, 0x5, 0x5, 0xc00010dd00, 0x0, 0x0)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/cmd/controller-gen/main.go:345 +0x27c
main.main.func1(0xc0001a0000, 0xc000170550, 0x5, 0x5, 0x0, 0x0)
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/cmd/controller-gen/main.go:217 +0x4a
github.com/spf13/cobra.(*Command).execute(0xc0001a0000, 0xc0000920d0, 0x5, 0x5, 0xc0001a0000, 0xc0000920d0)
	/Users/abursavich/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762 +0x465
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001a0000, 0x17b97c0, 0x397, 0x178e6da)
	/Users/abursavich/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852 +0x2ec
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/abursavich/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
main.main()
	/Users/abursavich/go/pkg/mod/sigs.k8s.io/controller-tools@v0.2.0-beta.1/cmd/controller-gen/main.go:223 +0x238
make: *** [manifests] Error 2
```

--- a/pkg/markers/collect.go
+++ b/pkg/markers/collect.go
@@ -290,7 +290,9 @@ func (v markerSubVisitor) Visit(node ast.Node) ast.Visitor {
 	}
 
 	// skip comments on the same line as the previous node
-	if v.lastLineCommentGroup != nil && v.lastLineCommentGroup.Pos() == v.allComments[v.commentInd].Pos() {
+	// making sure to double-check for the case where we've gone past the end of the comments
+	// but still have to finish up typespec-gendecl association (see below).
+	if v.lastLineCommentGroup != nil && v.commentInd < len(v.allComments) && v.lastLineCommentGroup.Pos() == v.allComments[v.commentInd].Pos() {
 		v.commentInd++
 	}
 

--- a/pkg/markers/markers_suite_test.go
+++ b/pkg/markers/markers_suite_test.go
@@ -54,7 +54,7 @@ var _ = BeforeSuite(func() {
 						foo "fmt" // +testing:pkglvl="not here import 3"
 
 						// +testing:pkglvl="not here import 4"
-					) 
+					)
 
 					// +testing:pkglvl="here unattached"
 
@@ -105,6 +105,15 @@ var _ = BeforeSuite(func() {
 
 					// +testing:typelvl="here on typedecl with no more"
 					type Cheese struct { }
+
+					// ensure that we're fine if we've got an end-of-line
+					// comment that's the last comment of the file, but
+					// we still have a bit more to traverse (field list --> ident).
+					// THIS MUST CONTAIN THE LAST COMMENT IN THE FILE
+					// TODO(directxman12): split this off into its own case
+					type private struct {
+						bar int // not collected
+					}
 				`,
 			},
 		},

